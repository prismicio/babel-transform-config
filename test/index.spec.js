const transformConfig = require('../');
const consola = require('consola');

describe('transformConfig.transform', () => {
  const config = `export default {
    foo: ['default'],
    foo2: ['default'],
    bar: {
      baz: ['default'],
      baz2: ['default']
    },
    qux: [{}]
  };`;

  /**
   * Creates
   */
  it('creates array', () => {
    const transforms = {
      test: {
        action: 'create',
        value: ['created']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('creates multiple arrays', () => {
    const transforms = {
      test: {
        action: 'create',
        value: ['created']
      },
      test2: {
        action: 'create',
        value: ['created2']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('creates array in object', () => {
    const transforms = {
      'bar:nested': {
        action: 'create',
        value: ['created']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('creates multiple arrays in object', () => {
    const transforms = {
      'bar:nested': {
        action: 'create',
        value: ['created']
      },
      'bar:nested2': {
        action: 'create',
        value: ['created2']
      },
      'bar:nested3': {
        action: 'create',
        value: ['created3']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('creates nested array', () => {
    const transforms = {
      'test:nested': {
        action: 'create',
        value: ['created']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('creates nested multiple arrays', () => {
    const transforms = {
      'test:nested': {
        action: 'create',
        value: ['created']
      },
      'test:nested2': {
        action: 'create',
        value: ['created2']
      },
      'test:nested3': {
        action: 'create',
        value: ['created3']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });

  /**
   * Replaces
   */
  it('replaces array', () => {
    const transforms = {
      foo: {
        action: 'replace',
        value: ['replaced']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('replaces multiple arrays', () => {
    const transforms = {
      foo: {
        action: 'replace',
        value: ['replaced']
      },
      foo2: {
        action: 'replace',
        value: ['replaced2']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('replaces array in object', () => {
    const transforms = {
      'bar:baz': {
        action: 'replace',
        value: ['replaced']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('replaces multiple arrays in object', () => {
    const transforms = {
      'bar:baz': {
        action: 'replace',
        value: ['replaced']
      },
      'bar:baz2': {
        action: 'replace',
        value: ['replaced2']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });

  /**
   * Merges
   */
  it('merges array', () => {
    const transforms = {
      foo: {
        action: 'merge',
        value: ['merged']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('merges multiple arrays', () => {
    const transforms = {
      foo: {
        action: 'merge',
        value: ['merged']
      },
      foo2: {
        action: 'merge',
        value: ['merged2']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('merges array in object', () => {
    const transforms = {
      'bar:baz': {
        action: 'merge',
        value: ['merged']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('merges multiple arrays in object', () => {
    const transforms = {
      'bar:baz': {
        action: 'merge',
        value: ['merged']
      },
      'bar:baz2': {
        action: 'merge',
        value: ['merged2']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });

  /**
   * Deletes
   */
  it('deletes array', () => {
    const transforms = {
      foo: {
        action: 'delete'
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('deletes multiple arrays', () => {
    const transforms = {
      foo: {
        action: 'delete'
      },
      foo2: {
        action: 'delete'
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('deletes array in object', () => {
    const transforms = {
      'bar:baz': {
        action: 'delete'
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('deletes multiple arrays in object', () => {
    const transforms = {
      'bar:baz': {
        action: 'delete'
      },
      'bar:baz2': {
        action: 'delete'
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
  it('deletes object', () => {
    const transforms = {
      bar: {
        action: 'delete'
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });

  /**
   * Creates/Replaces
   */
  it('creates/replaces array', () => {
    const transformsArray = [
      {
        foo: {
          action: 'create:replace',
          value: ['created/replaced']
        }
      },
      {
        test: {
          action: 'create:replace',
          value: ['created/replaced']
        }
      }
    ];
    const transformed = transformsArray.map(transforms =>
      transformConfig.transform(config, transforms)
    );
    expect(transformed.map(t => t.code)).toMatchSnapshot();
  });
  it('creates/replaces array in object', () => {
    const transformsArray = [
      {
        'bar:baz': {
          action: 'create:replace',
          value: ['created/replaced']
        }
      },
      {
        'bar:nested': {
          action: 'create:replace',
          value: ['created/replaced']
        }
      }
    ];
    const transformed = transformsArray.map(transforms =>
      transformConfig.transform(config, transforms)
    );
    expect(transformed.map(t => t.code)).toMatchSnapshot();
  });
  it('creates/replaces nested array', () => {
    const transforms = {
      'test:nested': {
        action: 'create:replace',
        value: ['created/replaced']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });

  /**
   * Creates/Merges
   */
  it('creates/merges array', () => {
    const transformsArray = [
      {
        foo: {
          action: 'create:merge',
          value: ['created/merged']
        }
      },
      {
        test: {
          action: 'create:merge',
          value: ['created/merged']
        }
      }
    ];
    const transformed = transformsArray.map(transforms =>
      transformConfig.transform(config, transforms)
    );
    expect(transformed.map(t => t.code)).toMatchSnapshot();
  });
  it('creates/merges array in object', () => {
    const transformsArray = [
      {
        'bar:baz': {
          action: 'create:merge',
          value: ['created/merged']
        }
      },
      {
        'bar:nested': {
          action: 'create:merge',
          value: ['created/merged']
        }
      }
    ];
    const transformed = transformsArray.map(transforms =>
      transformConfig.transform(config, transforms)
    );
    expect(transformed.map(t => t.code)).toMatchSnapshot();
  });
  it('creates/merges nested array', () => {
    const transforms = {
      'test:nested': {
        action: 'create:merge',
        value: ['created/merged']
      }
    };
    const transformed = transformConfig.transform(config, transforms);
    expect(transformed.code).toMatchSnapshot();
  });
});

describe('transformConfig', () => {
  it('should call consola.error when no transforms are made', () => {
    consola.mockTypes(typeName => typeName === 'error' && jest.fn());

    const result = transformConfig('', 'nuxt', {});

    expect(consola.error).toBeCalledWith('No transforms performed');
  });

  it('should call consola.error when no transforms are made', () => {
    consola.mockTypes(typeName => typeName === 'error' && jest.fn());
    const framework = 'foo';
    const message = `[transform-configs] ${framework} Framework not supported\nUse babel plugin directly instead`;

    const result = transformConfig('', framework, {});

    expect(consola.error).toBeCalledWith(message);
  });

  it('should call consola.error when no transforms are made', () => {
    consola.mockTypes(typeName => typeName === 'error' && jest.fn());
    const framework = '__proto__';
    const message = `[transform-configs] ${framework} Framework not supported\nUse babel plugin directly instead`;

    const result = transformConfig('', framework, {});

    expect(consola.error).toBeCalledWith(message);
  });
});
