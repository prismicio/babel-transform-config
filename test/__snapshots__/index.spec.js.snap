// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformConfig.transform creates array 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  },
  test: [\\"created\\"]
};"
`;

exports[`transformConfig.transform creates array in object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default'],
    nested: [\\"created\\"]
  }
};"
`;

exports[`transformConfig.transform creates multiple arrays 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  },
  test: [\\"created\\"],
  test2: [\\"created2\\"]
};"
`;

exports[`transformConfig.transform creates multiple arrays in object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default'],
    nested: [\\"created\\"],
    nested2: [\\"created2\\"]
  }
};"
`;

exports[`transformConfig.transform creates nested array 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  },
  test: {
    nested: [\\"created\\"]
  }
};"
`;

exports[`transformConfig.transform creates nested multiple arrays 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  },
  test: {
    nested: [\\"created\\"],
    nested3: [\\"created2\\"]
  }
};"
`;

exports[`transformConfig.transform creates/merges array 1`] = `
Array [
  "export default {
  foo: ['default', \\"created/merged\\"],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  }
};",
  "export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  },
  test: [\\"created/merged\\"]
};",
]
`;

exports[`transformConfig.transform creates/merges array in object 1`] = `
Array [
  "export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default', \\"created/merged\\"],
    baz2: ['default']
  }
};",
  "export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default'],
    nested: [\\"created/merged\\"]
  }
};",
]
`;

exports[`transformConfig.transform creates/merges nested array 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  },
  test: {
    nested: [\\"created/merged\\"]
  }
};"
`;

exports[`transformConfig.transform creates/replaces array 1`] = `
Array [
  "export default {
  foo: [\\"created/replaced\\"],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  }
};",
  "export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  },
  test: [\\"created/replaced\\"]
};",
]
`;

exports[`transformConfig.transform creates/replaces array in object 1`] = `
Array [
  "export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: [\\"created/replaced\\"],
    baz2: ['default']
  }
};",
  "export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default'],
    nested: [\\"created/replaced\\"]
  }
};",
]
`;

exports[`transformConfig.transform creates/replaces nested array 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  },
  test: {
    nested: [\\"created/replaced\\"]
  }
};"
`;

exports[`transformConfig.transform deletes array 1`] = `
"export default {
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform deletes array in object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform deletes multiple arrays 1`] = `
"export default {
  bar: {
    baz: ['default'],
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform deletes multiple arrays in object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {}
};"
`;

exports[`transformConfig.transform deletes object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default']
};"
`;

exports[`transformConfig.transform merges array 1`] = `
"export default {
  foo: ['default', \\"merged\\"],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform merges array in object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default', \\"merged\\"],
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform merges multiple arrays 1`] = `
"export default {
  foo: ['default', \\"merged\\"],
  foo2: ['default', \\"merged2\\"],
  bar: {
    baz: ['default'],
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform merges multiple arrays in object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: ['default', \\"merged\\"],
    baz2: ['default', \\"merged2\\"]
  }
};"
`;

exports[`transformConfig.transform replaces array 1`] = `
"export default {
  foo: [\\"replaced\\"],
  foo2: ['default'],
  bar: {
    baz: ['default'],
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform replaces array in object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: [\\"replaced\\"],
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform replaces multiple arrays 1`] = `
"export default {
  foo: [\\"replaced\\"],
  foo2: [\\"replaced2\\"],
  bar: {
    baz: ['default'],
    baz2: ['default']
  }
};"
`;

exports[`transformConfig.transform replaces multiple arrays in object 1`] = `
"export default {
  foo: ['default'],
  foo2: ['default'],
  bar: {
    baz: [\\"replaced\\"],
    baz2: [\\"replaced2\\"]
  }
};"
`;
